### PR TITLE
[FR] Add support for sample base queries in elasticsearch

### PR DESCRIPTION
**Pull request by Mikaayenson**
## Issues
Resolves https://github.com/endgameinc/eql/issues/70

## Details
- Adds support for sample base query types for Elasticsearch when `allow_sample` config is specified. 
- Exposes allow_runs to ensure the feature isn't used before it was introduced

## Testing

Created a sample query and stream against a list of events. Tested with and without joins. Expect to return `AnalyticOutput` hits.

### CODE REVIEW

**Review by eric-forte-elastic — COMMENTED** [review]
Adding potentially correct type hints (please modify as needed).

File: eql/ast.py [hunk 1]
```diff
@@ -944,6 +944,32 @@ def _render(self):
...
+    def __init__(self, queries, join_keys=None):
+        """Create a Sample of multiple events.
+
+        :param list[SubqueryBy] queries: List of queries to be sampled
+        :param list[str] join_keys: List of join keys
+        """
+        self.queries = queries
+        self.join_keys = join_keys or []
+
+    def _render(self):
```
[the remark this answers is not in the document]
Comment (Mikaayenson) [reply eql/ast.py:964]: We have to support py2.7 as well.

[the remark this answers is not in the document]
Comment (eric-forte-elastic) [reply eql/ast.py:964]: Ah yup good call thanks :+1:

**Review by eric-forte-elastic — APPROVED** [review]
LGTM :+1:

**Review by terrancedejesus — APPROVED** [review]
Looks good. No suggestions.

File: eql/ast.py [hunk 2]
```diff
@@ -944,6 +944,32 @@ def _render(self):
...
+        text = 'sample'
+        if self.join_keys:
+            text += ' by ' + ', '.join(self.join_keys)
+        text += '\n'
+        text += self.indent('\n'.join(query.render() for query in self.queries))
```
Comment (Mikaayenson) [inline eql/ast.py:969]: Calling query.render() adds the sample's join keys to each subquery. Not sure if this is correct.

Comment (Mikaayenson) [reply eql/ast.py:964]: https://github.com/endgameinc/eql/pull/74/commits/711fb45dde009bee4cfcb1ce66233409e7b7b057

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,14 @@
 # Event Query Language - Changelog
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+# Version 0.9.17
```
Comment (Mikaayenson) [inline CHANGELOG.md:4]:
```suggestion
# Version 0.9.18
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,14 @@
 # Event Query Language - Changelog
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+# Version 0.9.18
+
+ _Released 2023-08-20_
```
Comment (Mikaayenson) [inline CHANGELOG.md:6]: TODO: Will update before merging

**Comment by brokensound77** [comment]
- add to [highlighters](https://github.com/endgameinc/eql/blob/1b262f83d1c920c0bb1aef82be4104d408442783/eql/highlighters.py#L35)
- add tests validating the engine; looks like you will need to add `sample`-specific support there

**Comment by terrancedejesus** [comment]
✅ Took some time on 08-28-2023 to do a peer review with @Mikaayenson and @eric-forte-elastic. We reviewed the `_convert_sample_term` and `_convert_sample` methods in the engine itself. Example queries and events were used, as shown below, to simulate the functionality of this. It was concluded that the functionality works as expected based on analytic and subquery results for the sample feature.

Additionally, unit tests attempting to identify if a query can be parsed and loaded as an analytic was discussed, which seems to function correctly.

```
    def test_engine_load(self):
        """Check that various queries can be converted and loaded into the python engine."""
        engine = PythonEngine()
        engine.add_custom_function('myFn', lambda x, y, z: 100)


        sample_queries = [
            'sample by host.mac [generic where other.field: "test"] [generic where other.field: "test2"]',
            # 'sample [generic where other.field: "test"] [generic where other.field: "test"]',
            # 'sample [generic where other.field: "test"] [generic where other.field: "test2"]',
            # 'sample by x,y,z [event where a == 1] [event where b == 2]',
            # 'sample [event where name == "test"] [event where name == "test"]'
        ]

        with ignore_missing_functions, allow_sample, elasticsearch_syntax:
            for query in sample_queries:
                results = []
                # Make sure every query can be converted without raising any exceptions
                parsed_query = parse_query(query)
                engine.add_query(parsed_query)
                engine.add_output_hook(results.append)

                # Also try to load it as an analytic
                # parsed_analytic = parse_analytic({'metadata': {'id': uuid.uuid4()}, 'query': query})
                # engine.add_analytic(parsed_analytic)
                engine.stream_events([{"host": {"mac": ["42:01:0a:0c:00:6a"]}, "other": {"field": "test"}}, {"host": {"mac": ["42:01:0a:0c:00:6a"]}, "other": {"field": "test2"}}])
                print(results)
```

**Comment by Mikaayenson** [comment]
# Update Sep 1

Based on the manual detection rule [testing](https://github.com/elastic/detection-rules/pull/3000#issuecomment-1703058735) and unit testing within this PR, this is finally ready to go!

**Pull request merged**